tests/topotests/bgp_bmp/test_bgp_bmp.py

```diff
@@ -0,0 +1,243 @@
+    text_output = tgen.gears["bmp1"].run("cat /var/log/bmp-192_0_2_1.log")
+
+    for m in text_output.splitlines():
+        # some output in the bash can break the message deconding
+        try:
+            messages.append(json.loads(m))
+        except Exception as e:
+            logger.warning(str(e) + " message: {}".format(str(m)))
+            continue
+
+    if not messages:
+        logger.error("Bad BMP log format, check your BMP server")
+
+    return messages
+
+
+def check_for_prefixes(expected_prefixes, msg_type, policy):
+    """
+    Check for the presence of the given prefixes in the BMP server logs with
+    the given message type and the set policy.
+    """
+    global SEQ
+    # we care only about the new messages
+    messages = [
+        m for m in sorted(get_bmp_messages(), key=lambda d: d["seq"])
+        if m["seq"] > SEQ
+    ]
+
+    # get the list of pairs (prefix, policy, seq) for the given message type
+    p_flag = "is_post" if policy == "is_post" else "is_in"
+    prefix_policy = [
+        (m["ip_prefix"], m[p_flag], m["seq"]) for m in messages
+        if "log_type" in m.keys() and m["log_type"] == msg_type
+    ]
+
+    # check for prefixes
+    prefixes = [p[0] for p in prefix_policy]
+    for ep in expected_prefixes:
+        msg = "The {}ed Prefix {} is not present in the log messages."
+        assert ep in prefixes, msg.format(msg_type, ep)
+
+    # check for policy flag, (policy, seq)
+    policies = [(p[1], p[2]) for p in prefix_policy]
+    for pf in policies:
+        msg = "The BMP message {} has not the right message policy type {}"
+        assert int(pf[0]) == 1, msg.format(pf[1], policy)
+
+    SEQ = messages[-1]["seq"]
+
+
+def ipv4_v6_unicast_prefixes(policy):
+    """
+    Setup the BMP  monitor policy, Add and withdraw ipv4/v6 prefixes.
+    Check if the previous actions are logged in the BMP server with the right
+    message type and the right policy.
+    """
+    tgen = get_topogen()
+    # configure the bmp policy on the r1 router
+    del_policy = "post-policy" if policy == "pre-policy" else "pre-policy"
+    cmd = [
+        "con t\n",
+        "router bgp 65501\n",
+        "bmp targets bmp1\n",
+        "bmp monitor ipv4 unicast {}\n".format(policy),
+        "bmp monitor ipv6 unicast {}\n".format(policy),
+        "no bmp monitor ipv4 unicast {}\n".format(del_policy),
+        "no bmp monitor ipv6 unicast {}\n".format(del_policy),
+        "end\n",
+    ]
+    topotest.sleep(3, "setting BMP {}...".format(policy))
+    tgen.gears["r1"].vtysh_cmd("".join(cmd))
+
+    # add prefixes on r2
+    cmd = [
+        "conf t\n",
+        "router bgp 65502\n"
+        "address-family ipv4 unicast\n"
+        "network 172.31.0.15/32\n"
+        "exit-address-family\n"
+        "address-family ipv6 unicast\n"
+        "network 2001::1111/128\n"
+        "exit-address-family\n",
+    ]
+
+    tgen.gears["r2"].vtysh_cmd("".join(cmd))
+    topotest.sleep(2, "adding prefixes...")
```

Comment:
Same

Reply:
Sleep is needed here,  so the logs could be communicated to the bmp server before making the check.

Reply:
To clarify why sleep is not good.  Effectively we have just made a guess that 2 seconds is enough time.  Under heavy system load this will not work.  In fact we've had time after time where someone put a sleep in, the test vm's were under extremely heavy load( for whatever reason's, but not limited to other tests running in parallel ) and the sleep was insufficient.

The strategy that must be employed is loop (see run_and_expect()) that looks for the correct answer over a period of time that is guaranteed to be large enough for the answer to propagate.

Reply:
Thank you for the clarification. I went with run_and_expect().